src/acp/util/getAgentAccess.ts

```diff
@@ -0,0 +1,147 @@
+  // internal_getActorAccess in acp_v2:96 doesn't reduce the policies properly
+  // policyAppliesTo in acp_v2:256 assumes that every matcher is an agent matcher
+  //
+  // TODO: Stopgap solution
+  // Implement a simplistic reduce function that
+  // matches policies where the agent is present in the matchers
+  const allOfMatchers = getUrlAll(policy, ACP.allOf)
+    .map((url) => getThing(internal_getAcr(acr), url))
+    .filter((thing): thing is ThingPersisted => thing !== null);
+
+  const allOfMatched = allOfMatchers.every((thing) => {
+    return getUrlAll(thing, ACP.agent).includes(webId);
+  });
+
+  const anyOfMatchers = getUrlAll(policy, ACP.anyOf)
+    .map((url) => getThing(internal_getAcr(acr), url))
+    .filter((thing): thing is ThingPersisted => thing !== null);
+
+  const anyOfMatched = anyOfMatchers.some((thing) => {
+    return getUrlAll(thing, ACP.agent).includes(webId);
+  });
+
+  const noneOfMatchers = getUrlAll(policy, ACP.noneOf)
+    .map((url) => getThing(internal_getAcr(acr), url))
+    .filter((thing): thing is ThingPersisted => thing !== null);
+
+  const noneOfMatched = noneOfMatchers.some((thing) => {
+    return getUrlAll(thing, ACP.agent).includes(webId);
+  });
+
+  return (
+    (allOfMatchers.length === 0 || allOfMatched) &&
+    (anyOfMatchers.length === 0 || anyOfMatched) &&
+    (noneOfMatchers.length === 0 || !noneOfMatched)
+  );
+}
+
+/** @hidden */
+function reduceModes(
+  policy: ThingPersisted,
+  modes: AccessModes,
+  type: "control" | "resource"
+): AccessModes {
+  const allowed = getAllowModes(policy);
+  const denied = getDenyModes(policy);
+
+  if (type === "control") {
+    return {
+      read: modes.read,
+      append: modes.append,
+      write: modes.write,
+      controlRead: (modes.controlRead || allowed.read) && !denied.read,
+      controlWrite: (modes.controlWrite || allowed.write) && !denied.write,
+    };
+  }
+  return {
+    read: (modes.read || allowed.read) && !denied.read,
+    append: (modes.append || allowed.append) && !denied.append,
+    write: (modes.write || allowed.write) && !denied.write,
+    controlRead: modes.controlRead,
+    controlWrite: modes.controlWrite,
+  };
+}
+
+/**
+ * Get an overview of what access is defined for a given Agent.
+ *
+ * @param resourceWithAcr URL of the Resource you want to read the access for.
+ * @param webId WebID of the Agent you want to get the access for.
+ * @param options Default Options such as a fetch function.
+ * @since 1.16.0
+ */
+export async function getAgentAccess(
+  resourceWithAcr: WithAccessibleAcr,
+  webId: WebId
+): Promise<AccessModes> {
+  // TODO: add support for external resources
+  let resourceAccess = {
+    read: false,
+    append: false,
+    write: false,
+    controlRead: false,
+    controlWrite: false,
+  };
+
+  const policyAll = getPolicyUrlAll(resourceWithAcr)
+    .map((url) => getThing(internal_getAcr(resourceWithAcr), url))
+    .filter((policy): policy is ThingPersisted => policy !== null);
```

Comment:
Interesting, I didn't realize you could add type guards inline like this.

Reply:
:)
Yes, actually on my projects I have among my type guards toolbox this one:

```typescript
function isNonNullable<T>(x: T): x is NonNullable<T> {
  return typeof(x) !== "undefined" && x !== null;
}
```

Then you can do:
```typescript
    .filter(isNonNullable);
```

hehehe

Reply:
TypeScript is beautiful! ;)